### ISSUE TITLE
StackOverflowError due to subclassing

### ISSUE THREAD

**Issue by dts-ait** [mbknor/mbknor-jackson-jsonSchema#9]
I am trying to create a schema for a base class and class extending this base class with an additional attribute. (De)serialization with Jackson works fine, but when I try to generate a schema I get a StackOverflowError.

Am I missing something / doing something illegal in my implementation or is this a bug?

### BaseClass.java
```java
package schemaproblem;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({ @JsonSubTypes.Type(value = BaseClass.class, name = "BaseClass"),
		@JsonSubTypes.Type(value = ExtendedClass.class, name = "ExtendedClass") })
@JsonInclude(Include.NON_EMPTY)
public class BaseClass<T extends BaseClass<T>> {

	private int myInteger;

	public int getMyInteger() {
		return myInteger;
	}

	public T setMyInteger(int myInteger) {
		this.myInteger = myInteger;
		return (T) this;
	}

	@Override
	public String toString() {
		return "BaseClass [myInteger=" + myInteger + "]";
	}

}
```

### ExtendedClass.java
```java
package schemaproblem;

public class ExtendedClass extends BaseClass<ExtendedClass> {

	private String anotherDetail;

	public String getAnotherDetail() {
		return anotherDetail;
	}

	public ExtendedClass setAnotherDetail(String anotherDetail) {
		this.anotherDetail = anotherDetail;
		return this;
	}

	@Override
	public String toString() {
		return super.toString() + " -> ExtendedClass [anotherDetail=" + anotherDetail + "]";
	}
}
```

### SchemaExample.java

```java
package schemaproblem;

import java.io.File;
import java.io.IOException;

import com.fasterxml.jackson.core.JsonGenerationException;
import com.fasterxml.jackson.core.JsonParseException;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.kjetland.jackson.jsonSchema.JsonSchemaGenerator;

import scala.Option;

public class SchemaExample {

	public static final String exampleFile = "/tmp/example.json";
	private ObjectMapper mapper;

	public SchemaExample() {
		mapper = new ObjectMapper();
		mapper.findAndRegisterModules();
		mapper.enable(SerializationFeature.INDENT_OUTPUT);
	}

	public static void main(String[] args) throws JsonGenerationException, JsonMappingException, IOException {
		SchemaExample main = new SchemaExample();
		main.readAndWriteBaseClassJson();
		main.readAndWriteExtendedJson();
		main.writeSchemav4();
	}

	public void readAndWriteBaseClassJson() throws JsonParseException, JsonMappingException, IOException {
		BaseClass<?> example = new BaseClass<>().setMyInteger(42);
		mapper.writeValue(new File(exampleFile), example);
		BaseClass<?> readExample = mapper.readValue(new File(exampleFile), BaseClass.class);
		System.out.println("read from json file: " + readExample);
	}

	public void readAndWriteExtendedJson() throws JsonGenerationException, JsonMappingException, IOException {
		ExtendedClass example = new ExtendedClass().setMyInteger(42).setAnotherDetail("hello world");
		mapper.writeValue(new File(exampleFile), example);
		ExtendedClass readExample = mapper.readValue(new File(exampleFile), ExtendedClass.class);
		System.out.println("read from json file: " + readExample);
	}

	public void writeSchemav4() throws JsonGenerationException, IOException {
		JsonSchemaGenerator jsonSchemaGenerator = new JsonSchemaGenerator(mapper);
		Option<String> emtpyOption = Option.empty();
		JsonNode jsonSchema = jsonSchemaGenerator.generateJsonSchema(BaseClass.class, emtpyOption, emtpyOption);
		System.out.println("schema:\n" + mapper.writeValueAsString(jsonSchema));
	}

}
```

### The output

```
read from json file: BaseClass [myInteger=42]
read from json file: BaseClass [myInteger=42] -> ExtendedClass [anotherDetail=hello world]
Exception in thread "main" java.lang.StackOverflowError
	at scala.collection.mutable.ArrayBuilder.<init>(ArrayBuilder.scala:22)
	at scala.collection.mutable.ArrayBuilder$ofRef.<init>(ArrayBuilder.scala:56)
	at scala.collection.mutable.ArrayBuilder$.make(ArrayBuilder.scala:47)
	at scala.Array$$anon$2.apply(Array.scala:63)
	at scala.Array$$anon$2.apply(Array.scala:62)
	at scala.collection.TraversableLike$class.builder$1(TraversableLike.scala:229)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:233)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$5.apply(JsonSchemaGenerator.scala:447)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$5.apply(JsonSchemaGenerator.scala:447)
	at scala.Option.map(Option.scala:146)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper.expectObjectFormat(JsonSchemaGenerator.scala:446)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.acceptJsonFormatVisitor(BeanSerializerBase.java:809)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.acceptJsonFormatVisitor(DefaultSerializerProvider.java:580)
	at com.fasterxml.jackson.databind.ObjectMapper.acceptJsonFormatVisitor(ObjectMapper.java:3641)
	at com.fasterxml.jackson.databind.ObjectMapper.acceptJsonFormatVisitor(ObjectMapper.java:3620)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$expectObjectFormat$1$$anonfun$7.apply(JsonSchemaGenerator.scala:466)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$expectObjectFormat$1$$anonfun$7.apply(JsonSchemaGenerator.scala:463)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$DefinitionsHandler.getOrCreateDefinition(JsonSchemaGenerator.scala:160)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$expectObjectFormat$1.apply(JsonSchemaGenerator.scala:462)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$expectObjectFormat$1.apply(JsonSchemaGenerator.scala:459)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper.expectObjectFormat(JsonSchemaGenerator.scala:458)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.acceptJsonFormatVisitor(BeanSerializerBase.java:809)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.acceptJsonFormatVisitor(DefaultSerializerProvider.java:580)
	at com.fasterxml.jackson.databind.ObjectMapper.acceptJsonFormatVisitor(ObjectMapper.java:3641)
	at com.fasterxml.jackson.databind.ObjectMapper.acceptJsonFormatVisitor(ObjectMapper.java:3620)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$expectObjectFormat$1$$anonfun$7.apply(JsonSchemaGenerator.scala:466)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$expectObjectFormat$1$$anonfun$7.apply(JsonSchemaGenerator.scala:463)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$DefinitionsHandler.getOrCreateDefinition(JsonSchemaGenerator.scala:160)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$expectObjectFormat$1.apply(JsonSchemaGenerator.scala:462)
	at com.kjetland.jackson.jsonSchema.JsonSchemaGenerator$MyJsonFormatVisitorWrapper$$anonfun$expectObjectFormat$1.apply(JsonSchemaGenerator.scala:459)
	at scala.collection.immutable.List.foreach(List.scala:381)

[...]
```

**Comment by mbknor** [comment]
This might have been fixed by #6  in master, but it has not been released yet.

Can you clone this repo, build it and retry and see if the same problem exists?

**Comment by mbknor** [comment]
1.0.8 has now been released and should be available in maven central soon

**Comment by dts-ait** [comment]
Great, the issue is resolved with 1.0.8 :+1:! Maybe you can use my example as test case?
